src/Abc.Zebus.Directory.RocksDb/Storage/RocksDbPeerRepository.cs

```diff
@@ -0,0 +1,342 @@
+                                                     peer.IsResponding,
+                                                     peerDescriptor.IsPersistent,
+                                                     peerDescriptor.TimestampUtc.GetValueOrDefault(),
+                                                     peerDescriptor.HasDebuggerAttached,
+                                                     bindingKeys);
+            var destination = new MemoryStream();
+            Serializer.Serialize(destination, storagePeer);
+            var value = destination.ToArray();
+
+            _db.Put(key, value, _peersColumnFamily);
+        }
+
+        public PeerDescriptor Get(PeerId peerId)
+        {
+            var peerStorageBytes = _db.Get(BuildPeerKey(peerId), _peersColumnFamily);
+            if (peerStorageBytes == null)
+                return null;
+
+            return DeserialisePeerDescriptor(peerId, peerStorageBytes, GetDynamicSubscriptions(peerId));
+        }
+
+        private static PeerDescriptor DeserialisePeerDescriptor(PeerId peerId, byte[] peerStorageBytes, ICollection<Subscription> dynamicSubscriptions = null)
+        {
+            var peerStorage = Serializer.Deserialize<RocksDbStoragePeer>(new MemoryStream(peerStorageBytes));
+            var staticSubscriptions = peerStorage.StaticSubscriptions ?? Array.Empty<Subscription>();
+            var subscriptions = staticSubscriptions.Concat(dynamicSubscriptions ?? Array.Empty<Subscription>()).Distinct();
+            return new PeerDescriptor(peerId, peerStorage.EndPoint, peerStorage.IsPersistent, peerStorage.IsUp, peerStorage.IsResponding, peerStorage.TimestampUtc, subscriptions.ToArray())
+            {
+                HasDebuggerAttached = peerStorage.HasDebuggerAttached,
+            };
+        }
+
+        private Subscription[] GetDynamicSubscriptions(PeerId peerId)
+        {
+            var bindingKeys = new List<(MessageTypeId, BindingKey[])>();
+            IterateDynamicSubscriptionsForPeer(peerId,
+                                               (key, value) =>
+                                               {
+                                                   var (bindingKey, messageTypeId) = ReadDynamicSubscription(key, value);
+                                                   bindingKeys.Add((messageTypeId, bindingKey));
+                                               });
+            return bindingKeys.SelectMany(y => y.Item2.Select(bindingKey => new Subscription(y.Item1, bindingKey))).ToArray();
+        }
+
+        private void IterateDynamicSubscriptionsForPeer(PeerId peerId, Action<byte[], byte[]> action)
+        {
+            var key = BuildSubscriptionKeyPrefix(peerId);
+            using var cursor = _db.NewIterator(_subscriptionsColumnFamily);
+            if (!cursor.Seek(key).Valid())
+                return;
+
+            var peerIdLength = peerId.ToString().Length;
+            byte[] currentKey;
+            do
```

Comment:
The prefix does not seem to be verified for the first key... are we sure the behavior will be correct if we have two peer ids with a common prefix? I mean for instance if you're looking for `Foo` but the db contains `FooBar` you could get the `FooBar` value.

Likewise, `CompareStart` does not seem to handle this case.

Maybe the key structure should be `PeerId | \0 | MessageFullTypeName` so we could use `PeerId | \0` as the prefix to test?

Reply:
You're right we either need to add a separator or check the peer id length encoded in the key.